### PR TITLE
feat(pkg/triedb): Add trie access recorder

### LINKED ISSUE #3999: (pkg/trie): Implement TrieDB recorder
In order to implement the merkle proof using our TrieDB, we have to implement the recorder structure / feature.
This will record the path as we are traversing the trie, simplifying the proof generation process.

We can take some inspiration from:
https://github.com/paritytech/trie/blob/master/trie-db/src/recorder.rs

### PR DESCRIPTION
**Pull request by dimartiro**
## Changes

Add trieDB recorder to record all not inlined visited nodes during lookups.
This is the first step to be able to generate merkle proofs using the new `TrieDB`

### Summary of changes

- [x] Introduce trie `Recorder`
    - [x] Tests
- [x] Modify `TrieDB` constructor to receive variable opts

## Tests

```sh
make test
```

## Issues

closes: #3999

### CODE REVIEW

**Review by EclesioMeloJunior — APPROVED** [review]
LGTM!

File: pkg/trie/triedb/recorder.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	InlineValueAccess struct {
+		fullKey []byte
+	}
+	HashAccess struct {
+		fullKey []byte
+	}
+	NonExistingNodeAccess struct {
+		fullKey []byte
+	}
+)
```
Comment (timwu20) [inline pkg/trie/triedb/recorder.go]: do these need to be public?

File: pkg/trie/triedb/recorder.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	// We are not using common.Hash here since hash size could be > 32 bytes when we use prefixed keys.
+	// See ValueAccess.hash
+	hash []byte
+	data []byte
+}
+
+type Recorder struct {
+	nodes        []Record
+	recordedKeys btree.Map[string, RecordedForKey]
+}
```
Comment (timwu20) [inline pkg/trie/triedb/recorder.go]: Same for these, do they need to be public?

File: pkg/trie/triedb/recorder.go (hunk 1 above)
Comment (dimartiro) [reply pkg/trie/triedb/recorder.go]: Noup, fixed

File: pkg/trie/triedb/recorder.go (hunk 2 above)
Comment (dimartiro) [reply pkg/trie/triedb/recorder.go]: Yes, since they are going to be used by the proof generation in a different package

**Review by timwu20 — APPROVED** [review]

**Pull request merged**